tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+			merr.Add(f.Close())
+			returnErr = merr.Err()
+		}
+	}()
+
+	cdm.size += cdm.n
+	atomic.StoreInt64(&cdm.n, int64(n))
+	oldSeq := cdm.curFileSequence
+	cdm.curFileSequence = seq
+
+	// Write the provided time for mint and 0 for maxt.
+	cdm.curFileMint = mint
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(mint))
+	if _, err := f.Write(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return err
+	}
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], 0)
+	if _, err := f.Write(cdm.writeBuf[:MintMaxtSize]); err != nil {
+		return err
+	}
+	atomic.AddInt64(&cdm.n, 16)
+
+	oldFile := cdm.curFile
+
+	cdm.curFile = f
+	if cdm.wbuf != nil {
+		cdm.wbuf.Reset(f)
+	} else {
+		cdm.wbuf = bufio.NewWriterSize(f, writeBufferSize)
+	}
+
+	if oldFile != nil {
+		// Open it again with the new size.
+		newTailFile, err := fileutil.OpenMmapFile(oldFile.Name())
+		if err != nil {
+			return err
+		}
+		cdm.mmappedChunksMtx.Lock()
+		oldMmapFile := cdm.closers[oldSeq]
+		cdm.closers[oldSeq] = newTailFile
+		cdm.mmappedChunkFiles[oldSeq] = realByteSlice(newTailFile.Bytes())
+		cdm.mmappedChunksMtx.Unlock()
+		// Closing the last mmapped file.
+		if err := oldMmapFile.Close(); err != nil {
+			cdm.mmappedChunksMtx.Unlock()
+			return err
+		}
+	}
+
+	mmapFile, err := fileutil.OpenMmapFileWithSize(f.Name(), int(MaxHeadChunkFileSize))
+	if err != nil {
+		return err
+	}
+	cdm.mmappedChunksMtx.Lock()
+	cdm.closers[cdm.curFileSequence] = mmapFile
+	cdm.mmappedChunkFiles[cdm.curFileSequence] = realByteSlice(mmapFile.Bytes())
+	cdm.mmappedChunksMtx.Unlock()
+
+	return nil
+}
+
+// finalizeCurFile writes all pending data to the current tail file,
+// truncates its size, and closes it.
+func (cdm *ChunkDiskMapper) finalizeCurFile() error {
+	if cdm.curFile == nil {
+		return nil
+	}
+
+	if err := cdm.flushBuffer(); err != nil {
+		return err
+	}
+
+	binary.BigEndian.PutUint64(cdm.writeBuf[:], uint64(cdm.curFileMaxt))
+	if _, err := cdm.curFile.WriteAt(cdm.writeBuf[:MintMaxtSize], HeaderMaxtOffset); err != nil {
+		return nil
+	}
+
+	if err := cdm.curFile.Sync(); err != nil {
+		return err
+	}
+
+	return cdm.curFile.Close()
+}
+
+func (cdm *ChunkDiskMapper) write(b []byte) error {
+	n, err := cdm.wbuf.Write(b)
+	atomic.AddInt64(&cdm.n, int64(n))
+	return err
+}
+
+func (cdm *ChunkDiskMapper) writeWithCRC32(b []byte) error {
+	if err := cdm.write(b); err != nil {
+		return err
+	}
+	_, err := cdm.crc32.Write(b)
+	return err
+}
+
+// flushBuffer flushes the current in-memory chunks.
+// Assumes that wbufLock is _write_ locked before calling this method.
+func (cdm *ChunkDiskMapper) flushBuffer() error {
+	if err := cdm.wbuf.Flush(); err != nil {
+		return err
+	}
+	for i := 0; i < inBufferShards; i++ {
+		cdm.inBufferChunksLocks[i].Lock()
+		cdm.inBufferChunks[i] = make(map[uint64]chunkenc.Chunk)
+		cdm.inBufferChunksLocks[i].Unlock()
+	}
+	return nil
+}
+
+// Chunk returns a chunk from a given reference.
+func (cdm *ChunkDiskMapper) Chunk(ref uint64) (chunkenc.Chunk, error) {
```

Comment:
Note that the returned chunk's bytes are not available after Close of ChunkDiskMapper maybe?

Reply:
Will add that in the comment.